File: flink-staging/flink-language-binding/flink-python/src/main/java/org/apache/flink/languagebinding/api/java/python/io/PythonInputFormat.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.languagebinding.api.java.python.io;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import org.apache.flink.api.common.typeinfo.TypeInformation;
import org.apache.flink.api.common.io.FileInputFormat;
import org.apache.flink.api.java.tuple.Tuple;
import org.apache.flink.api.java.typeutils.ResultTypeQueryable;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.core.fs.FileInputSplit;
import org.apache.flink.core.fs.FileStatus;
import org.apache.flink.core.fs.Path;
import org.apache.flink.languagebinding.api.java.python.streaming.PythonStreamer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 */
public class PythonInputFormat<T extends Tuple> extends FileInputFormat<T> implements ResultTypeQueryable<T> {

	//exclusively needed for executing split logic in python
	private final PythonStreamer splitStreamer;
	private PythonStreamer streamer;
	private boolean streamerOpen;
	private boolean readForSplit;
	static final Logger LOG = LoggerFactory.getLogger(PythonInputFormat.class);
	private TypeInformation<T> typeInformation;
	private Path path;
	private Configuration configuration;
	private PythonCollector<T> collector;
	protected int readRecords = 0;
	private String filter;
	private boolean splitsInPython;

	public void setTypeInformation(TypeInformation<T> typeInformation) {
		this.typeInformation = typeInformation;
	}

	public PythonInputFormat(Path path, int id, TypeInformation<T> info, String filter, boolean computeSplitsInPython) {
		super(path);
		this.path = path;
		this.filter = filter;
		this.streamer = new PythonStreamer(this, id);
		this.splitStreamer = new PythonStreamer(this, id, true);
		this.typeInformation = info;
		this.collector  = new PythonCollector<T>();
		this.splitsInPython = computeSplitsInPython;
	}

	@Override
	public FileInputSplit[] createInputSplits(int minNumSplits) throws IOException {
		if(this.splitsInPython) {
			//do fancy stuff in python here
			//TODO: make sure runtime context is not used
			//TODO: make special methods to execute this part on the JobManager
			this.splitStreamer.open();
			//TODO send computation request here
			//TODO: refactor sendCloseMessage to send Message
			this.splitStreamer.sendMessage("compute_splits");
			return null;
		}else {
			FileInputSplit[] inputSplits = super.createInputSplits(minNumSplits);
			System.out.println("break");
			return inputSplits;
		}
	}

	@Override
	protected boolean acceptFile(FileStatus fileStatus) {
		String name = fileStatus.getPath().getName();
		return name.matches(this.filter) && super.acceptFile(fileStatus);
	}

	@Override
	public T nextRecord(T record) throws IOException {
		if(!readForSplit) {
			readEnviTile();
			this.readForSplit = true;
		}

		if(!this.reachedEnd()) {
			record = this.collector.poll();
			this.readRecords++;
		}else {
			record = null;
		}
		return record;
	}
	
	@Override
	public void configure(Configuration parameters) {
		// TODO Auto-generated method stub
		super.configure(parameters);
		this.configuration = parameters;
	}

	private void readEnviTile() throws IOException {
		List<String> pathList = new ArrayList<String>();
		pathList.add(this.path.toString());
		this.streamer.streamBufferWithoutGroups(pathList.iterator(), collector);
	}

	@Override
	public TypeInformation<T> getProducedType() {
		return this.typeInformation;
	}
	
	@Override
	public void open(FileInputSplit split) throws IOException {
		// TODO Auto-generated method stub
		super.open(split);
		this.path = split.getPath();
		this.readForSplit = false;
		this.collector.clear();
		if (!this.streamerOpen) {
			this.streamer.open();
			this.streamer.sendMessage("open_task");
			this.streamer.sendBroadCastVariables(this.configuration);
			System.out.println("Java: sent BC vars");
			this.streamerOpen = true;
		}

	}

	@Override
	public void close() throws IOException {
		super.close();
	}

	@Override
	public boolean reachedEnd() throws IOException {
		return this.collector.isEmpty() && this.readForSplit;
	}

	public void destroy() throws Exception{
		if(this.streamerOpen) {
			this.streamer.sendMessage("close");
			this.streamer.close();
		}
		super.destroy();
	}
}

File: flink-libraries/flink-python/src/main/python/org/apache/flink/python/api/flink/io/PythonInputFormat.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
################################################################################
from flink.connection import Connection, Iterator, Collector
from flink.functions import RuntimeContext, Function
import sys


class FileInputSplit(object):
    def __init__(self, path, start, end, hosts):
        self.path = path
        self.start = start
        self.end = end
        self.hosts = hosts


class PythonInputFormat(Function.Function):
    def __init__(self):
        super(PythonInputFormat, self).__init__()
        self.close_called = False

    def _run(self):
        self._iterator._setLargeTuples(False)
        collector = self._collector
        function = self.deliver
        print("before iterator next")
        sys.stdout.flush()
        split = self._iterator.next()
        print("after iterator next")
        print(split)
        print(type(split))
        print(type("close"))
        sys.stdout.flush()

        if split[0] == "close":
            self.close_called = True
        while split is not None and split[0] != "close":
            try:
                function(split, collector)
            except:
                print("in function call:", sys.exc_info()[0])
            try:
                self._iterator._reset()
            except:
                print("in iterator reset call:", sys.exc_info()[0])
                sys.stdout.flush()
            try:
                self._connection.send_end_signal()
            except:
                print("in connection end call:", sys.exc_info()[0])
                sys.stdout.flush()
            try:
                split = self._iterator.next()
            except:
                print("in iterator next call:" + str(self.close_called), sys.exc_info()[0])
                sys.stdout.flush()
            if split[0] == "close":
                print("close split encountered")
                sys.stdout.flush()
                self.close_called = True


        collector._close()

    def deliver(self, path, collector):
        pass

    def computeSplits(self, env, con):
        iterator = Iterator.PlanIterator(con, env)
        collector = Collector.SplitCollector(con, env)

        min_num_splits = iterator.next()
        path = iterator.next()

        self.createInputSplits(min_num_splits, path, collector)

        collector._close()
    
    def createInputSplits(self, minNumSplits, path, collector):
        pass
